### PR TITLE
[vulkan] inplace add_, relu_

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42181 [vulkan] Permtute ops: transpose,view,slice,unsqueeze,select
* #42180 [vulkan][op] avg_pool2d
* #42179 [vulkan][op] add.Scalar, mul.Scalar
* #41434 [vulkan] cat op (concatenate)
* **#41380 [vulkan] inplace add_, relu_**

Differential Revision: [D22754939](https://our.internmc.facebook.com/intern/diff/D22754939)

### CODE REVIEW

File: aten/src/ATen/native/vulkan/VulkanAten.cpp [hunk 1]
```diff
@@ -272,6 +272,18 @@ Tensor vulkan_add(const Tensor& self, const Tensor& other, const Scalar alpha) {
   return new_with_vtensor_vulkan(std::move(output), self.options());
 }
 
+Tensor& vulkan_add_(Tensor& self, const Tensor& other, Scalar alpha) {
+  auto& x = vtensor_from_vulkan(self.is_vulkan() ? self : self.vulkan());
+  auto& y = vtensor_from_vulkan(other.is_vulkan() ? other : other.vulkan());
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/vulkan/VulkanAten.cpp:277]: const.  I think vtensor_from_vulkan should be overloaded on const.

File: aten/src/ATen/native/vulkan/VulkanAten.cpp [hunk 2]
```diff
@@ -272,6 +272,18 @@ Tensor vulkan_add(const Tensor& self, const Tensor& other, const Scalar alpha) {
...
+  float a = alpha.to<float>();
+
+  VulkanTensor output{self.sizes().vec()};
+  output.allocate_storage();
+  vulkan::detail::add(output, x, y, a);
+  x = std::move(output);
+  return self;
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/vulkan/VulkanAten.cpp:284]: I'm not sure I understand this logic here.  If self is already allocated and the goal of this function is to implement c += a * b, then do we need to allocate output as well?  Can we not do it in place in 'self'?  Also, why the move?  Can't we just pass output, output, and input and a (in that order) to vulkan::detail::add?

File: aten/src/ATen/native/vulkan/VulkanAten.cpp [hunk 3]
```diff
@@ -413,23 +425,30 @@ Tensor vulkan_clamp(
...
-  self.copy_(y);
+  auto& x = vtensor_from_vulkan(self);
+  VulkanTensor output{self.sizes().vec()};
+  output.allocate_storage();
+  vulkan::detail::clamp(
+      output,
+      x,
+      min ? min.value().to<float>() : -std::numeric_limits<float>::infinity(),
+      max ? max.value().to<float>() : std::numeric_limits<float>::infinity());
+  x = std::move(output);
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/vulkan/VulkanAten.cpp:451]: Same.  If this operation is in place, then why do we need to `allocate_storage`?

File: aten/src/ATen/native/vulkan/VulkanAten.cpp (hunk 2 above)
Comment (IvanKobzarev) [reply aten/src/ATen/native/vulkan/VulkanAten.cpp:281]: Here we read from x (self) and y (other) and write output values to output.  
At the moment all shaders work having separate output and input(s), here it is using the same shader that is used for copy variant of this op.

`x = std::move(output)` replaces x (self) content (pointer to VulkanTensor::Impl) of x to the result of output.

`vulkan::detail::add()` uses only `image()` representation, so we do not really need allocated buffer(). 
For this implementation we need `output.allocate_storage()`, but with changes  https://github.com/pytorch/pytorch/pull/42569 allocate_storage can be removed

File: aten/src/ATen/native/vulkan/VulkanAten.cpp (hunk 1 above)
Comment (IvanKobzarev) [reply aten/src/ATen/native/vulkan/VulkanAten.cpp]: I will add const override for vtensor_from_vulkan to this PR

File: aten/src/ATen/native/vulkan/VulkanAten.cpp (hunk 3 above)
Comment (IvanKobzarev) [reply aten/src/ATen/native/vulkan/VulkanAten.cpp:472]: (As in previous comment) Output VulkanTensor is used here as a temporary variable to reuse shaders with separate output, inputs; allocate_storage to remove in https://github.com/pytorch/pytorch/pull/42569 :)

**Review by AshkanAliabadi — APPROVED** [review]

**Pull request closed without merging**